vortex-io/src/aligned.rs

```diff
@@ -0,0 +1,161 @@
+#![allow(dead_code)]
+use std::ops::{Deref, DerefMut};
+
+use bytes::Bytes;
+
+pub trait PowerOfTwo<const N: usize> {}
+impl<const N: usize> PowerOfTwo<N> for usize where usize: sealed::Sealed<N> {}
+
+mod sealed {
+    pub trait Sealed<const N: usize> {}
+
+    impl Sealed<1> for usize {}
+    impl Sealed<2> for usize {}
+    impl Sealed<4> for usize {}
+    impl Sealed<8> for usize {}
+    impl Sealed<16> for usize {}
+    impl Sealed<32> for usize {}
+    impl Sealed<64> for usize {}
+    impl Sealed<128> for usize {}
+    impl Sealed<256> for usize {}
+    impl Sealed<512> for usize {}
+}
+
+/// A variant of [`BytesMut`][bytes::BytesMut] that freezes into a [`Bytes`] that is guaranteed
+/// to begin at a multiple of a target byte-alignment.
+///
+/// Internally, it accomplishes this by over-allocating by up to the alignment size, padding the
+/// front as necessary. Reads and writes will only be able to access the region after the padding.
+///
+/// It is required for the alignment to be a valid power of 2 <= 512, any other value will be
+/// a compile-time failure.
+pub(crate) struct AlignedBytesMut<const ALIGN: usize> {
+    buf: Vec<u8>,
+    padding: usize,
+    capacity: usize,
+}
+
+impl<const ALIGN: usize> AlignedBytesMut<ALIGN>
+where
+    usize: PowerOfTwo<ALIGN>,
+{
+    /// Allocate a new mutable buffer with capacity to hold at least `capacity` bytes.
+    ///
+    /// The mutable buffer may allocate more than  the requested amount to pad the memory for
+    /// alignment.
+    pub fn with_capacity(capacity: usize) -> Self {
+        // Allocate up to `ALIGN` extra bytes, in case we need to pad the returned pointer.
+        let allocation_size = (capacity + ALIGN - 1).next_multiple_of(ALIGN);
+        let mut buf = Vec::<u8>::with_capacity(allocation_size);
+        let padding = buf.as_ptr().align_offset(ALIGN);
+        unsafe {
+            buf.set_len(padding);
+        }
+
+        Self {
+            buf,
+            padding,
+            capacity,
+        }
+    }
+
+    /// Usable capacity of this buffer.
+    pub fn capacity(&self) -> usize {
+        self.capacity
+    }
+
+    /// Set the length of the mutable buffer directly.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for ensuring that the provided length fits within the original
+    /// capacity request.
+    ///
+    /// Failure to do so could cause uninitialized memory to be readable.
+    pub unsafe fn set_len(&mut self, len: usize) {
+        unsafe { self.buf.set_len(len + self.padding) }
```

Comment:
consider adding `debug_assert!(len + self.padding <= self.capacity)` or even plain old `assert!`, I doubt we'll mind the performance penalty of the check

Reply:
I was considering this but I saw that `Vec::set_len` doesn't do this check.

However, it looks like `BytesMut` does

<img width="625" alt="image" src="https://github.com/user-attachments/assets/6a830241-8540-4401-9067-ba07d76e6889">

I can add it

Reply:
cool, went ahead and approved, you should merge after you add the assertion